modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java

```diff
@@ -1801,4 +1842,178 @@ private boolean enterBusy() {
     private void leaveBusy() {
         opsLock.readLock().unlock();
     }
+
+    /**
+     * Registers metrics to measure durations of service methods.
+     *
+     * @param srvc Service for invocations measurement.
+     * @param srvcName Name of {@code srvc}.
+     */
+    private void registerMetrics(Service srvc, String srvcName) {
+        getInterfaces(srvc.getClass()).stream().map(Class::getMethods).flatMap(Arrays::stream)
```

Comment:
@ivandasch , can you clarify what you mean 'thread-safe'? Isn't 'computeIfAbsent' is not enough? I suggest to make sure there is no concurrent call of `IgniteProcessor.undeploy()`:

```
            synchronized (ctxs) {
                if(!ctxs.iterator().next().isCancelled())
                    registerMetrics(srvc, srvcCtx.name());
            }
```

WDYT?

Reply:
Well, lets take a look over the javaDoc
```
The default implementation makes no guarantees about synchronization or atomicity properties of this method. Any implementation providing atomicity guarantees must override this method and document its concurrency properties. In particular, all implementations of subinterface java.util.concurrent.ConcurrentMap must document whether the function is applied once atomically only if the value is not present.
```

Reply:
Same here. You are right about usage of separate method. But the synchronization you suggest seems already relies in the logic above. I don’t think `IgniteServiceProcessor.undeploy()` and ` IgniteServiceProcessor.redeploy()` can happen concurrently but we should consider it can, ok. Let’s see:
-	Metrics are erased by `IgniteServiceProcessor.unregisterMetrics()` in 2 places (): `IgniteServiceProcessor.redeploy()` and `IgniteServiceProcessor.undeploy()`. Both leads to removing whole map by service name ‘invocationHistograms.remove(srvcName);’
-	Metrics are created by `IgniteServiceProcessor.registerMetrics()`  in one place: `IgniteServiceProcessor.redeploy()`.

These calls are not any APIs, are private.

All 3 calls covered with sync by service context collection given bt same serviced:
```
Collection<ServiceContextImpl> ctxs …
synchronized (ctxs) {
	unregisterMetrics(cfg.getName());
        }
```
and
```
Collection<ServiceContextImpl> ctxs …
            synchronized (ctxs) {
                if (!ctxs.iterator().next().isCancelled())
                    registerMetrics(srvc, srvcCtx.name());
            } 
```

I think I really missed the last part earlier. Thanks for the note. Isn’t it the synchronized block you suggest?